naps/cores/cmv12k/s7_phy.py

```diff
@@ -0,0 +1,472 @@
+            self.ctrl_lane_delay_inc = 2 # increment clock delay
+            delay -= 1
+
+    @driver_method
+    def configure_sensor(self, sensor_spi):
+        # set default train pattern
+        sensor_spi.set_train_pattern(0b101010_010101)
+        self.lane_pattern = 0b101010_010101
+        # switch (only) sensor sequencer to 12 bit mode
+        sensor_spi.set_bit_mode(12)
+
+    @driver_method
+    def initial_alignment(self):
+        # try all possible options to align the control lane
+
+        for half in range(2):
+            for slip in range(6):
+                self.ctrl_lane_delay_reset = 1 # zero control lane delay
+                for delay in range(32):
+                    if self.ctrl_lane_match:
+                        return True
+                    self.ctrl_lane_delay_inc = 1 # bump control lane delay
+                self.ctrl_lane_bitslip = 1 # slip control lane by one bit
+            self.ctrl_lane_bitslip = 2 # slip data half words
+
+        return False
+
+    @driver_method
+    def data_alignment(self):
+        import bitarray.util
+        num_lanes = self.num_lanes
+
+        # determine the best bitslip as the one which matches over the most delays
+        best_bitslip = [0]*(num_lanes+1)
+        best_bitslip_matches = [0]*(num_lanes+1)
+        for bitslip in range(6):
+            match_count = [0]*(num_lanes+1)
+            self.data_lane_delay_reset = (1<<num_lanes)-1 # zero all lane delays
+            self.ctrl_lane_delay_reset = 1
+            for lane_delay in range(32):
+                all_matches = self.data_lane_match | (self.ctrl_lane_match << num_lanes)
+                for lane in range(num_lanes+1):
+                    match_count[lane] += 1 if all_matches & (1 << lane) else 0
+                self.data_lane_delay_inc = (1<<num_lanes)-1 # increment all delays
+                self.ctrl_lane_delay_inc = 1
+            for lane in range(num_lanes+1):
+                if best_bitslip_matches[lane] < match_count[lane]:
+                    best_bitslip_matches[lane] = match_count[lane]
+                    best_bitslip[lane] = bitslip
+            self.data_lane_bitslip = (1<<num_lanes)-1 # slip all bits
+            self.ctrl_lane_bitslip = 1
+
+        # apply the determined bitslip to all the channels
+        for bitslip in range(6):
+            apply_mask = 0
+            for lane in range(num_lanes):
+                if best_bitslip[lane] > 0:
+                    apply_mask |= (1 << lane)
+                    best_bitslip[lane] -= 1
+            self.data_lane_bitslip = apply_mask
+            if best_bitslip[num_lanes] > 0:
+                self.ctrl_lane_bitslip = 1
+                best_bitslip[num_lanes] -= 1
+
+        # measure the delay window for all the channels to determine the final delay.
+        # the window has a 1 bit for each delay tap value which matches properly
+        delay_window = list(bitarray.util.zeros(32) for _ in range(num_lanes+1))
```

Comment:
Might be cleaner as a list comprehension.

Reply:
I agree.